### PR TITLE
136 intuitive frontend interface for pipeline definition

### PR DESCRIPTION
**Pull request by Gaetan-cogliati**
Interactive pipeline creation feature

### CODE REVIEW

File: frontend/src/components/CardBase/CardBase.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import React, {ReactNode} from "react";
+import {Tag} from "../../models/Tag";
+
+const CardBase : React.FC<{
+    index: number, item: any, tags: Tag[], handleTags: any, cardActions: ReactNode, cardName: ReactNode, summary: ReactNode
+}> = ({index, item, tags, handleTags, cardActions, cardName, summary}) => {
+
+
+    return (
+        <Grid xs={12} sm={6} lg={4} xl={3} key={index}
```
Comment (andrptrc) [inline frontend/src/components/CardBase/CardBase.tsx:15]: If the component is called CardBase, it should only be about a Card item, not a Grid of Cards, maybe use CardBase in a ItemGridBase or something similar

File: frontend/src/enums/dataTypeEnum.tsx [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    TEXT_CSV = "text/csv",
+    APPLICATION_JSON = "application/json",
+    APPLICATION_PDF = "application/pdf",
+    AUDIO_MP3 = "audio/mpeg",
+    AUDIO_OGG = "audio/ogg"
+}
+
+ export const DataTypeOptions: string[] = [];
+
+Object.values(DataType).forEach((value) => {
```
Comment (andrptrc) [inline frontend/src/enums/dataTypeEnum.tsx:15]: Why do we need to add this to an array to export ?

File: frontend/src/enums/dataTypeEnum.tsx [hunk 3]
```diff
@@ -0,0 +1,17 @@
+// each tag has an acronym, name, and a unique set of colors
```
Comment (andrptrc) [inline frontend/src/enums/dataTypeEnum.tsx:1]: Is this comment correct ?

**Review by andrptrc — CHANGES_REQUESTED** [review]
Generally, try to lint/format the code when you write it (Cmd+opt+L on mac in WebStorm) so it is easier to read. Will discuss with you about some other improvements in the next meeting

File: frontend/src/components/ServiceCard/HomeServiceCard.tsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    index: number, item: any, tags: Tag[], handleTags: any, ai: boolean, handleAIToggle: any, functions: any
+}> = ({index, item, tags, handleTags, ai, handleAIToggle, functions}) => {
+
+
+    return (
+        <ServiceCardBase index={index} item={item} tags={tags} handleTags={handleTags} ai={ai}
+                         handleAIToggle={handleAIToggle}>
+            <Link
+                to={"/showcase/service/" + item.slug}>
+                <Button size={"small"} variant={"contained"}>View</Button>
```
Comment (andrptrc) [inline frontend/src/components/ServiceCard/HomeServiceCard.tsx:20]: add `disableElevation` to button (basically do it on every button that variant is not outlined

File: frontend/src/components/ItemGrid/PipelineList.tsx [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                                            <Typography>
+                                                {item.summary}
+                                            </Typography>
+                                        </CardContent>
+                                        <CardActions sx={{p: 2}}>
+                                            <Link
+                                                to={"/showcase/pipeline/" + item.slug}
+                                                state={{back: searchParams.toString()}}
+                                            >
+                                                <Button size={"small"} variant={"contained"}>View</Button>
```
Comment (andrptrc) [inline frontend/src/components/ItemGrid/PipelineList.tsx:76]: Also add `disableElevation` here

File: frontend/src/components/Edges/EditEdge.tsx
Comment (andrptrc) [inline frontend/src/components/Edges/EditEdge.tsx:1]: Add background or another way to improve visibility of condition. Perhaps putting the edge z index on top of everything else could improve this

File: frontend/src/components/ItemGrid/PipelineList.tsx
Comment (andrptrc) [inline frontend/src/components/ItemGrid/PipelineList.tsx:1]: Why is there a pipelineList but no serviceList component ? Is this file used ? Maybe adapt it and create the service one and include them in the code of other components

File: frontend/src/enums/dataTypeEnum.tsx [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    IMAGE_PNG = "image/png",
+    TEXT_PLAIN = "text/plain",
+    TEXT_CSV = "text/csv",
+    APPLICATION_JSON = "application/json",
+    APPLICATION_PDF = "application/pdf",
+    AUDIO_MP3 = "audio/mpeg",
+    AUDIO_OGG = "audio/ogg"
+}
+
+ export const DataTypeOptions: string[] = [];
```
Comment (andrptrc) [inline frontend/src/enums/dataTypeEnum.tsx:12]: maybe change uppercase first letter to lower case

File: frontend/src/pages/Showcase/index.tsx [hunk 7]
```diff
@@ -147,3 +147,5 @@ const Showcase: React.FC<{ mobileOpen: boolean }> = ({mobileOpen}) => {
 }
 
 export default Showcase;
+
```
Comment (andrptrc) [inline frontend/src/pages/Showcase/index.tsx:150]: Useless blank lines

File: frontend/src/utils/api.tsx [hunk 8]
```diff
@@ -97,6 +97,66 @@ export const getPipelines = async (filter: string, skip: number, limit: number,
...
 }
 
+
+export const checkPipelineValidity = async (jsonDescription: string)=> {
+    /*
+     * Function used to check if a pipeline is valid (using a Json representation of said pipeline)
+     * jsonDescription: string - the pipeline's description, in json format.
+     */
+
+    const headers = {
```
Comment (andrptrc) [inline frontend/src/utils/api.tsx:107]: Are these headers mandatory ? or could we use the const variable on top of the file ?

File: frontend/src/utils/api.tsx [hunk 9]
```diff
@@ -97,6 +97,66 @@ export const getPipelines = async (filter: string, skip: number, limit: number,
...
+
+}
+
+export const createPipeline = async (jsonDescription: string)=> {
+    /*
+     * Function used to check if a pipeline is valid (using a Json representation of said pipeline)
+     * jsonDescription: string - the pipeline's description, in json format.
+     */
+    const headers = {
+        'Access-Control-Allow-Origin': '*',
```
Comment (andrptrc) [inline frontend/src/utils/api.tsx:137]: Same here

File: frontend/src/pages/CreatePipeline/index.tsx [hunk 10]
```diff
@@ -0,0 +1,686 @@
...
+        })
+    }
+
+
+    React.useEffect(() => {
+        const initialNodes: Node<any>[] = []
+        initialNodes.push(createEntryNode())
+        initialNodes.push(createExitNode())
+        setNodes(() => initialNodes);
+    }, []);
```
Comment (andrptrc) [inline frontend/src/pages/CreatePipeline/index.tsx:574]: Either add dependencies to the array or remove it, but empty arrays should be avoided

File: frontend/src/pages/CreatePipeline/index.tsx [hunk 11]
```diff
@@ -0,0 +1,686 @@
...
+        const entryNode = nodesRef.current.find((node) => node.type === "entryNodeEdit");
+        const exitNode = nodesRef.current.find((node) => node.type === "exitNodeEdit");
+        const lastEdge = edgesRef.current.find((edge) => edge.target === exitNode?.id);
+
+        const data_in_fields = entryNode?.data.dataIn;
+        const data_out_fields = exitNode?.data.dataOut;
+        const queue: string[] = [lastEdge!.source]
+        let index = 0
+        while (index < queue.length) {
+            const node = nodesRef.current.find((nd) => nd.id === queue[index])
```
Comment (andrptrc) [inline frontend/src/pages/CreatePipeline/index.tsx:511]: ESLint: Function declared in a loop contains unsafe references to variable(s) 'index'.(no-loop-func)

File: frontend/src/pages/CreatePipeline/index.tsx [hunk 12]
```diff
@@ -0,0 +1,686 @@
...
+            position: {x: 550, y: 200},
+        }
+    }
+
+    const addServiceNode = (serviceName: string, serviceId: string, tags: any[], serviceSlug: string, dataIn: FieldDescription[], dataOut: FieldDescription[]) => {
+        let counter = 2;
+        const selectedDataIn = new Array<string>(dataIn.length);
+        let identifier = serviceSlug
+        let label = serviceName;
+        while (nodesRef.current.find((node) => node.data.identifier === identifier)) {
```
Comment (andrptrc) [inline frontend/src/pages/CreatePipeline/index.tsx:406]: ESLint: Function declared in a loop contains unsafe references to variable(s) 'index'.(no-loop-func)

File: frontend/src/pages/CreatePipeline/index.tsx [hunk 13]
```diff
@@ -0,0 +1,686 @@
...
+
+    const updateSelectedInputOption = (affectedNode: Node, newName: string, previousName: string) => {
+        const selectedDataIn = affectedNode.data.selectedDataIn.map((option: string) => {
+            if (option === previousName) return newName;
+            return option;
+        })
+        setNodeSelectedDataIn(affectedNode, selectedDataIn);
+    }
+
+    const onConnect = (params: Edge<any> | Connection) => {
```
Comment (andrptrc) [inline frontend/src/pages/CreatePipeline/index.tsx:176]: Explicit type argument any can be removed

**Review by andrptrc — CHANGES_REQUESTED** [review]
Also look at the following points:

- The file structure can be improved. For example put the cards in a `Cards` folder just like nodes, the point here is to be constant on the structure every where
- Check logs and remove every unused import or variable also checks for linting warnings
- Edit Nodes design could be improved to look a little more like the “non-editable” nodes
- Select-boxes style to match with other components and Board dimensions, settings, placing and spacing on page too
- Code improvements and bug fixing
- Code formatting (Cmd + alt + L on WebStorm)

**Comment by andrptrc** [comment]
Some more comments:

- How do we see the JSON result? Maybe a non-editable text-box like the one for the description on showcase page
- Try putting Service Cards in the sidebar
- When you link input to end -> error
- Unique key on each element (check error logs in browser
- Adapt the spacinf of fields for editing descriptions and other pipeline info to be the same as other parts of the website

**Comment by Gaetan-cogliati** [comment]
Elements to improve, add or fix:
- Find a way to let child nodes change the state of the 'nodes' and 'edges' state variables from the CreatePipeline page without having to use Refs (that need to be updated in a useEffect or useMemo).
- Put the service list in the sidebar
- Maybe restrict the selection of input types to compatible ones only, once the input handle is connected.
- Make sure the order of the pipeline inputs is correct in the generated JSON (pipeline description).
- Find a way to conveniently show the generated JSON to the user.

**Review by andrptrc — APPROVED** [review]

**Review by rmarquis — APPROVED** [review]

**Comment by rmarquis** [comment]
Looks okay to me.

I see some warning in the frontend section though. I don't know if these are related to this PR or if that's an older issue:

```
WARNING in [eslint]
frontend  | src/components/FilterDrawer/FilterDrawer.tsx
frontend  |   Line 63:11:  'MaterialUISwitch' is assigned a value but never used  @typescript-eslint/no-unused-vars
frontend  |   Line 99:11:  'CustomSwitch' is assigned a value but never used      @typescript-eslint/no-unused-vars
frontend  |
frontend  | src/pages/PipelineEditor/index.tsx
frontend  |   Line 614:8:  React Hook React.useEffect has a missing dependency: 'onEdgeDelete'. Either include it or remove the dependency array  react-hooks/exhaustive-deps
```

Also some deprecation warnings:
```
(node:37) [DEP_WEBPACK_DEV_SERVER_ON_AFTER_SETUP_MIDDLEWARE] DeprecationWarning: 'onAfterSetupMiddleware' option is deprecated. Please use the 'setupMiddlewares' option.
frontend  | (Use `node --trace-deprecation ...` to show where the warning was created)
frontend  | (node:37) [DEP_WEBPACK_DEV_SERVER_ON_BEFORE_SETUP_MIDDLEWARE] DeprecationWarning: 'onBeforeSetupMiddleware' option is deprecated. Please use the 'setupMiddlewares' option.
```

**Pull request merged**